### PR TITLE
Optimize GSU Workgroup Assignment

### PR DESCRIPTION
**Pull request by ssuyuanchang**
1. Workgroup order changed from interleaved to coalesced assignment.
2. Used "GlobalSplitUCoalesced" to enable in GSU kernel.
3. Add test yaml for "GlobalSplitUCoalesced" (increase ~500s test time)

### CODE REVIEW

**Review by KKyang — APPROVED** [review]

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -3805,10 +3844,26 @@ def calculateLoopNumIter(self, kernel, tPA, tPB, loopIdx):
...
+                                     comment="s%s = s[sgprSizesSum] / %s"%(tmpSgpr+1,depthU)))
+        module.add(scalarUInt32DivideAndRemainder(tmpSgpr, tmpSgpr+1, "GSU", "GSUSumIdx+1", tmpVgprRes, wavewidth=kernel["WavefrontSize"]))
+        self.vgprPool.checkIn(tmpVgpr)
+        module.add(SSubU32(dst=sgpr(tmpSgpr+1), src0=sgpr("GSU"), src1=1, comment="GSU-1"))
+        module.add(SCmpLgU32(src0=sgpr(tmpSgpr), src1=0, comment="quotient == 0"))
+        module.add(SCSelectB32(dst=sgpr(tmpSgpr), src0=sgpr("GSUSumIdx+1"), src1=sgpr(tmpSgpr+1), \
+                               comment="lastWg = (quotient==0) ? numIterPerWgRemainder : GSU-1"))
+        # if GSU numIter=0 if gsuSumIdx != lastWg
+        module.add(SCmpLgU32(src0=sgpr("GSUSumIdx"), src1=sgpr(tmpSgpr), comment="gsuSumIdx == lastWg"))
+        module.add(SCMovB32(dst=loopCounter, src=hex(0), comment="numIter=0 if gsuSimIdx != lastWg"))
```
Comment (jichangjichang) [inline tensilelite/Tensile/KernelWriterAssembly.py]: typo gsuSimIdx

Comment (ssuyuanchang) [reply tensilelite/Tensile/KernelWriterAssembly.py]: fixed

**Comment by ssuyuanchang** [comment]
Local cl passed results:
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------- generated xml file: /josh/hipBLASLt/tensilelite/python_tests.xml -------
[33m===== [32m51 passed[0m, [33m[1m18 skipped[0m, [33m[1m1 xfailed[0m, [33m[1m18 warnings[0m[33m in 21969.76s (6:06:09)[0m[33m =====[0m
py310: exit 0 (21970.07 seconds) /josh/hipBLASLt/tensilelite> py.test -v --basetemp=/josh/hipBLASLt/tensilelite/.tox/py310/tmp --junit-xml=/josh/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py310 --color=yes -n 4 --prebuilt-client=/josh/hipBLASLt/tensilelite/.tox/py310/client/0_Build/client/tensile_client Tensile/Tests -m common --capture=sys pid=20840
.pkg: _exit> python /usr/local/lib/python3.10/dist-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: exit None (0.00 seconds) /josh/hipBLASLt/tensilelite> python /usr/local/lib/python3.10/dist-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__ pid=12064
  py310: OK (22156.89=setup[21.09]+cmd[165.73,21970.07] seconds)
  congratulations :) (22156.98 seconds)

**Comment by hcman2** [comment]
> Local cl passed results: -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html ------- generated xml file: /josh/hipBLASLt/tensilelite/python_tests.xml ------- �[33m===== �[32m51 passed�[0m, �[33m�[1m18 skipped�[0m, �[33m�[1m1 xfailed�[0m, �[33m�[1m18 warnings�[0m�[33m in 21969.76s (6:06:09)�[0m�[33m =====�[0m py310: exit 0 (21970.07 seconds) /josh/hipBLASLt/tensilelite> py.test -v --basetemp=/josh/hipBLASLt/tensilelite/.tox/py310/tmp --junit-xml=/josh/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py310 --color=yes -n 4 --prebuilt-client=/josh/hipBLASLt/tensilelite/.tox/py310/client/0_Build/client/tensile_client Tensile/Tests -m common --capture=sys pid=20840 .pkg: _exit> python /usr/local/lib/python3.10/dist-packages/pyproject_api/_backend.py True setuptools.build_meta **legacy** .pkg: exit None (0.00 seconds) /josh/hipBLASLt/tensilelite> python /usr/local/lib/python3.10/dist-packages/pyproject_api/_backend.py True setuptools.build_meta **legacy** pid=12064 py310: OK (22156.89=setup[21.09]+cmd[165.73,21970.07] seconds) congratulations :) (22156.98 seconds)

**Comment by hcman2** [comment]
your ci pass only includes tensilelite. You need hipblaslt-test PASS

**Review by jichangjichang — DISMISSED** [review]
Please don't double size the tests of Tenisilelite. You should add your new arguments into some test which is focus on gsu

**Comment by ssuyuanchang** [comment]
image

**Comment by ssuyuanchang** [comment]
> Please don't double size the tests of Tenisilelite. You should add your new arguments into some test which is focus on gsu

Fixed by adding only two test yaml for this feature

**Comment by jichangjichang** [comment]
@ssuyuanchang  The different part between GSUC0 and GSUC1 is only in pre-loop global read address section and Tail loop section. Both of these sections are not in unroll-loop, which mean we can support these GSU0 and GSU1 in the same kernel. You can use a bit to distinguish which mode we want to use for each kernel launch.

@KKyang How do you think?

**Comment by ssuyuanchang** [comment]
> @ssuyuanchang The different part between GSUC0 and GSUC1 is only in pre-loop global read address section and Tail loop section. Both of these sections are not in unroll-loop, which mean we can support these GSU0 and GSU1 in the same kernel. You can use a bit to distinguish which mode we want to use for each kernel launch.
> 
> @KKyang How do you think?

How about merging this PR first and then combine GSUC0 and GSUC1 into the same kernel in the future changes?

**Comment by hcman2** [comment]
> @ssuyuanchang The different part between GSUC0 and GSUC1 is only in pre-loop global read address section and Tail loop section. Both of these sections are not in unroll-loop, which mean we can support these GSU0 and GSU1 in the same kernel. You can use a bit to distinguish which mode we want to use for each kernel launch.
> 
> @KKyang How do you think?

sounds like we can move many of the parameters outside the main loop by using mode config?

**Review by hcman2 — APPROVED** [review]
OK if the CI error is fixed after rebase.

**Pull request merged**